### ISSUE TITLE
[Bug]  Serialization/Deserialization concepts of Transaction

### ISSUE THREAD

**Issue by Neo-vortex** [bmresearch/Solnet#296]
**Describe the bug**
I think I have problem understanding the serialization/deserialization concepts of `Transaction`
I try to run the code below but the result is :
 `Transaction signature verification failure `

```
var tbuilder = new TransactionBuilder()
                .SetRecentBlockHash((await rpcClient.GetRecentBlockHashAsync()).Result.Value.Blockhash)
                .SetFeePayer(buyer.Account)
                .AddInstruction(MemoProgram.NewMemo(owner.PublicKey, "this is a test"));
            var signlaterwithmessage = tbuilder.CompileMessage();
            //this happens out of csharp code on the browser , but even doing it in csharp doesn't help as it doesn't work
            var clienttransacton = Transaction.Populate(Message.Deserialize(signlaterwithmessage)); 
            clienttransacton.Sign(buyer.Account);      
            var clientdto = clienttransacton.Serialize();
            //this happens in csharp code side
            var newtx = Transaction.Deserialize(clientdto);
            newtx.Sign(owner);
            RequestResult<string> tx2 = await rpcClient.SendTransactionAsync(newtx.Serialize());
            Console.WriteLine(tx2.Result);
            var subscription2 = await streamingRpcClient.SubscribeSignatureAsync(tx2.Result,
                (subscriptionState, response) => {
                }, Commitment.Finalized);
            while (true)
            {
                Console.WriteLine(tx2.Reason);  //Transaction signature verification failure 
                Console.WriteLine(subscription2.State);
                System.Threading.Thread.Sleep(1000);
            }
```

**Expected behavior**
I think it should work

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Solnet version 4.2 [as 4.17 doesn't work with Solana.Metaplex that we use]

**Comment by murlokito** [comment]
You are using the `Transaction` serialization/deserialization incorrectly. Check the examples [here](https://github.com/bmresearch/Solnet/blob/master/src/Solnet.Examples/TransactionDecodingExample.cs).

If you're using `Transaction.Deserialize()` it can be assumed the transaction is complete and there's no need to add signatures or to sign anything, the transaction is complete. 
`Transaction.Populate()` is used to populate a transaction with the message payload and a list of signatures from the necessary accounts, as you can see [here](https://github.com/bmresearch/Solnet/blob/master/src/Solnet.Examples/TransactionDecodingExample.cs#L80-L81).

**Comment by Neo-vortex** [comment]
Our desire is to communicate with Phantom wallet:
1-Phantom wallet has to sign the transaction, so we send  `Transaction.CompileMessage()` to the wallet
2- After that phantom wallet will signed the transaction with `window.solana.signTransaction(transaction)` [ [LINK TO DOC]](https://docs.phantom.app/integrating/sending-a-transaction) 
3-Now phantom wallet has to send the transaction with his signature back to server [here is the problem, how can I deserialization the transaction that Phantom send me ]
4-server can finally sign the  transaction himself  and then execute the transaction

There is a way that we tried that also failed. we tried to get the signature of the transaction that was signed by phantom wallet at step 2 and manually add it to transaction in  step 4. but did not work

Is there any standard protocol we must follow to get this working ?
As there seems to be no method to   deserialization  transaction from phantom wallet

**Comment by murlokito** [comment]
So this transaction has to be signed by the phantom wallet and another key pair in a server?

```
const signedTransaction = await window.solana.signTransaction(transaction);
const signature = await connection.sendRawTransaction(signedTransaction.serialize());
```

From what I understand the snippet above returns the `Transaction` and not the signature, so you could `Transaction.serializeMessage()` and encode the UInt8Array into base64 to send to the server, this would be the message that needs to be signed. You would also need to get the signature, probably from `signedTransaction.signatures` and send both to the server.

In the server you would need to do `var msg = Message.Deserialize(base64Message)`

And then you could finally do `Transaction.Populate(msg, new List<byte[]>{ serverAccount.sign(msg), signatureFromPhantom }`
The signatures you pass here into `Populate` need to be ordered according to which instruction requires the corresponding signer first, you must pay attention to this.

**Comment by nort3x** [comment]
i have a similar situation but it's reversed semantically, i will overview each steps need to be happen:

1- a complex transaction is being made with solnet (which i prefer to be done in server side)
2- i compile this transaction to a message which will be sent to client
3- client will populate a transaction and sign it from the given message ( it needs to be transaction not raw message cause extensions , like solfare and phantom don't show instructions when singing a message and just printed raw contents , bytes! , but with singing transaction they actually print out each step in human readable diagrams ) 

4- i have a singed transaction in client side which i need to be singed by server as well so i extract client signature and send it to server,

5- server sign the message itself , and it has both client publickey and signature of client as well,

6- server do `Transaction.populate(message, {clientSig,serverSig})` and generate a fully  ready to be sent transaction

7- i send it and i get `signature verification failure` 

although i can get expected results by just doing this:

1- message being made by server and sent to client 
2- client sign the `message` and send the signature to server
3- server sign the message and populate a transaction like above and it works!

so my question is :
1- is my procedure right? 
2- can i somehow stick to signing transaction on client only? because of privacy issues?
3- if message is stateless (cause solana docs says transaction is just a message with compact-array of signatures)
so i assume moving around message it self doesn't need to be happen, right?

**Comment by murlokito** [comment]
@nort3x before submitting the transaction on the server, when you populate it with the signatures, are these signatures in the correct order? The error you're describing also occurs when the signatures are out of order and they can't be verified

**Comment by murlokito** [comment]
I could try to replicate these issues you're having in order to provide a better solution and/or figure out if there's something that needs fixing but I would need more context, i.e what tech are you using for the app that signs with phantom? is this react app or something else? cuz if that's the case I don't think I could replicate these issues with our own Blazor Wasm dapp scaffold project

**Comment by nort3x** [comment]
yes the order got me couple of times but i'm sure it's fine now
i'm willing to help, but problem is i can't narrow the code down right now,
but i can add you to repositories so we can figure out the problem,

mean while i have a PoC created for this let me add it but it won't work out of the box but i will add some comments to clarify the problem

```csharp

           // a transaction which needs two signatures , client first, server second

            var TBuilder = new TransactionBuilder()
                .SetRecentBlockHash((await rpcClient.GetRecentBlockHashAsync()).Result.Value.Blockhash)
                .SetFeePayer(client.Account.PublicKey)
                ... some other instructions

            var message_as_bin  = TBuilder.CompileMessage();
            var what_i_expect  =  TBuilder.Build(new List<Account>() {client.Account, server.Account});
             


          // what need to be done

            // server signs
            Message message = Message.Deserialize(message_as_bin);
            var creatorSign = creater.Sign(message_as_bin);

           // client need to sign but transactions only
           Transaction t = Transaction.Populate(message);  // raw transaction
           t.Sign(client.Account); //it happens in browser
           t.AddSignature(creator.Account.PublicKey, creatorSign); // adding the signature of server

            Console.WriteLine("what we generate");
            Console.WriteLine(System.Text.Json.JsonSerializer.Serialize(t));
            Console.WriteLine("what we expect");
            Console.WriteLine(System.Text.Json.JsonSerializer.Serialize(Transaction.Deserialize(what_i_expect)));

```

basically i'm stuck in signing the transaction in multiple sessions and i have constraint of only signing transaction (not message) in one of the peers

**Comment by Neo-vortex** [comment]
> If you're using `Transaction.Deserialize()` it can be assumed the transaction is complete and there's no need to add signatures or to sign anything, the transaction is complete. 

The document is saying the opposite for `Transaction.Signatures` .
I quote the documentation:
`/// <summary>
    /// The signatures for the transaction.
    /// <remarks>
    /// These are typically created by invoking the <c>Build(IList{Account} signers)</c> method of the <see cref="T:Solnet.Rpc.Builders.TransactionBuilder" />,
    /// but can be created by deserializing a Transaction and adding signatures manually.
    /// </remarks>
    /// </summary>`

**Comment by Neo-vortex** [comment]
I think i may have narrowed down the problem a little bit
this examples don't work
```csharp
      var tbuilder1 = new TransactionBuilder()
                .SetRecentBlockHash((await rpcClient.GetRecentBlockHashAsync()).Result.Value.Blockhash)
                .SetFeePayer(buyer.Account.PublicKey)
                .AddInstruction(MemoProgram.NewMemo(creater.Account, "hello world"));
            var binmessge2  = tbuilder1.CompileMessage();
            var gg = Transaction.Populate(Message.Deserialize(binmessge2));
            gg.Sign(buyer.Account);
            gg.Sign(creater.Account);
 RequestResult<string> tx = await rpcClient.SendTransactionAsync(gg.Serialize());    //Transaction signature verification failure 
```
also this example doesn't work

```csharp
var tbuilder1 = new TransactionBuilder()
                .SetRecentBlockHash((await rpcClient.GetRecentBlockHashAsync()).Result.Value.Blockhash)
                .SetFeePayer(buyer.Account.PublicKey)
                .AddInstruction(MemoProgram.NewMemo(creater.Account, "hello world"));
            var binmessge2  = tbuilder1.CompileMessage();
            var gg = Transaction.Populate(Message.Deserialize(binmessge2));
            gg.Build(new List<Account>() {buyer.Account, creater.Account});
             RequestResult<string> tx = await rpcClient.SendTransactionAsync(gg.Serialize());    //Transaction signature verification failure 
```
only this one works
```csharp

     var tbuilder1 = new TransactionBuilder()
                .SetRecentBlockHash((await rpcClient.GetRecentBlockHashAsync()).Result.Value.Blockhash)
                .SetFeePayer(buyer.Account.PublicKey)
                .AddInstruction(MemoProgram.NewMemo(creater.Account, "hello world"));
               var binmessge2  = tbuilder1.CompileMessage();
              var gg = Transaction.Populate(Message.Deserialize(binmessge2));
             var binmessage3 = tbuilder1.Build(new List<Account>() {buyer.Account, creater.Account});
             RequestResult<string> tx = await rpcClient.SendTransactionAsync(binmessage3);      
```

**Comment by Neo-vortex** [comment]
I think i may have narrowed down the problem a little bit
this examples don't work
```csharp
      var tbuilder1 = new TransactionBuilder()
                .SetRecentBlockHash((await rpcClient.GetRecentBlockHashAsync()).Result.Value.Blockhash)
                .SetFeePayer(buyer.Account.PublicKey)
                .AddInstruction(MemoProgram.NewMemo(creater.Account, "hello world"));
            var binmessge2  = tbuilder1.CompileMessage();
            var gg = Transaction.Populate(Message.Deserialize(binmessge2));
            gg.Sign(buyer.Account);
            gg.Sign(creater.Account);
 RequestResult<string> tx = await rpcClient.SendTransactionAsync(gg.Serialize());    //Transaction signature verification failure 
```
also this example doesn't work

```csharp
var tbuilder1 = new TransactionBuilder()
                .SetRecentBlockHash((await rpcClient.GetRecentBlockHashAsync()).Result.Value.Blockhash)
                .SetFeePayer(buyer.Account.PublicKey)
                .AddInstruction(MemoProgram.NewMemo(creater.Account, "hello world"));
              var binmessge2  = tbuilder1.CompileMessage();
             var gg = Transaction.Populate(Message.Deserialize(binmessge2));
             var binmessage4 =   gg.Build(new List<Account>() {buyer.Account, creater.Account});
             RequestResult<string> tx = await rpcClient.SendTransactionAsync(binmessage4);    //Transaction signature verification failure 
```
only this one works
```csharp

     var tbuilder1 = new TransactionBuilder()
                .SetRecentBlockHash((await rpcClient.GetRecentBlockHashAsync()).Result.Value.Blockhash)
                .SetFeePayer(buyer.Account.PublicKey)
                .AddInstruction(MemoProgram.NewMemo(creater.Account, "hello world"));
               var binmessge2  = tbuilder1.CompileMessage();
              var gg = Transaction.Populate(Message.Deserialize(binmessge2));
             var binmessage3 = tbuilder1.Build(new List<Account>() {buyer.Account, creater.Account});
             RequestResult<string> tx = await rpcClient.SendTransactionAsync(binmessage3);      //OK
```

**Comment by nort3x** [comment]
any update?

**Comment by murlokito** [comment]
> any update?

I haven't had time to look more in depth into this as I've been working on something else that also uses these ser/des features for a multisig and offline signing wallet. Here's how I've been using these:

Crafting a transaction and serializing the message:

```csharp
NonceInformation nonceInfo = new NonceInformation()
{
    Nonce = NonceAccountViewModel.Nonce,
    Instruction = SystemProgram.AdvanceNonceAccount(
        new(NonceAccountViewModel.NonceAccountMap.Account),
        new(NonceAccountViewModel.NonceAccountMap.Authority)
    )
};
_txBuilder = new TransactionBuilder()
    .SetFeePayer(SourceAccount.PublicKey)
    .SetNonceInformation(nonceInfo);

// these calls only add instructions to the transaction
if (AccountContent is MultiSignatureAccountViewModel multiSigVm)
{
    CraftMultiSignatureTransaction(multiSigVm.SelectedSigners);
}
else
{
    CraftNonMultiSignatureTransaction();
}

_msgBytes = _txBuilder.CompileMessage();
Payload = Convert.ToBase64String(_msgBytes);
```

Signing the serialized message from above in a different context:

```csharp
var msgBytes = Convert.FromBase64String(Payload);
var signature = _walletService.CurrentWallet.Wallet.Account.Sign(msgBytes);
Signature = Convert.ToBase64String(signature);
```

And finally populating the transaction with the required signatures in another context:

```csharp
var msg = Message.Deserialize(Payload);
var tx = Transaction.Populate(msg,
    RequiredSignatures.Select(x => Convert.FromBase64String(x.Signature)).ToList());

Progress = "Submitting transaction..";
var txSig = await _rpcClient.SendTransactionAsync(tx.Serialize());
```

This code has been directly copied from the app and it's currently working.

**Comment by murlokito** [comment]
> yes the order got me couple of times but i'm sure it's fine now i'm willing to help, but problem is i can't narrow the code down right now, but i can add you to repositories so we can figure out the problem,
> 
> mean while i have a PoC created for this let me add it but it won't work out of the box but i will add some comments to clarify the problem
> 
> ```cs
>            // a transaction which needs two signatures , client first, server second
> 
>             var TBuilder = new TransactionBuilder()
>                 .SetRecentBlockHash((await rpcClient.GetRecentBlockHashAsync()).Result.Value.Blockhash)
>                 .SetFeePayer(client.Account.PublicKey)
>                 ... some other instructions
> 
>             var message_as_bin  = TBuilder.CompileMessage();
>             var what_i_expect  =  TBuilder.Build(new List<Account>() {client.Account, server.Account});
>              
> 
> 
>           // what need to be done
> 
>             // server signs
>             Message message = Message.Deserialize(message_as_bin);
>             var creatorSign = creater.Sign(message_as_bin);
> 
>            // client need to sign but transactions only
>            Transaction t = Transaction.Populate(message);  // raw transaction
>            t.Sign(client.Account); //it happens in browser
>            t.AddSignature(creator.Account.PublicKey, creatorSign); // adding the signature of server
> 
>             Console.WriteLine("what we generate");
>             Console.WriteLine(System.Text.Json.JsonSerializer.Serialize(t));
>             Console.WriteLine("what we expect");
>             Console.WriteLine(System.Text.Json.JsonSerializer.Serialize(Transaction.Deserialize(what_i_expect)));
> ```
> 
> basically i'm stuck in signing the transaction in multiple sessions and i have constraint of only signing transaction (not message) in one of the peers

Reading this again it looks like you are misunderstanding what a message and a transaction mean.
A message is the part of a transaction which specifies what happens when the transaction is executed by the runtime (the instructions it executes, the accounts they take, etc etc).
A transaction is a message along with the compact array of it's signatures.


When you do this:

// server signs
>             Message message = Message.Deserialize(message_as_bin);
>             var creatorSign = creater.Sign(message_as_bin);
> 
>            // client need to sign but transactions only
>            Transaction t = Transaction.Populate(message);  // raw transaction
>            t.Sign(client.Account); //it happens in browser
>            t.AddSignature(creator.Account.PublicKey, creatorSign); // adding the signature of server

You should be doing this (assuming the server assembles the transaction with the client signature)

```csharp
var message_as_bin  = TBuilder.CompileMessage(); // this is the wire format message
Message message = Message.Deserialize(message_as_bin); 
var creatorSign = creater.Sign(message_as_bin); // i assume this is the server signature

var clientSign = t.Sign(client.Account); // it happens in browser

Transaction t = Transaction.Populate(message, new List<byte[]>{ clientSign, creatorSign });
var txBytes = t.Serialize(); // this is the wire format transaction
```

**Comment by murlokito** [comment]
@nort3x @Neo-vortex Any updates on this since my last message? Need to figure out if something actually needs to be fixed.

**Comment by Neo-vortex** [comment]
I still don't understand why this example does not work

```csharp
  var tbuilder1 = new TransactionBuilder()
                .SetRecentBlockHash((await rpcClient.GetRecentBlockHashAsync()).Result.Value.Blockhash)
                .SetFeePayer(buyer.Account.PublicKey)
                .AddInstruction(MemoProgram.NewMemo(creater.Account, "hello world"));
            var binmessge2  = tbuilder1.CompileMessage();
            var gg = Transaction.Populate(Message.Deserialize(binmessge2));
            gg.Sign(buyer.Account);
            gg.Sign(creater.Account);
 RequestResult<string> tx = await rpcClient.SendTransactionAsync(gg.Serialize());    //Transaction signature verification failure 
```

As you said 'A transaction is a message along with the compact array of it's signatures.'.
when I populate a transaction from a message with no signature , it means i am creating a transaction with empty signature array.
so when i do `gg.Sign(buyer.Account);`, I am adding the signature of `buyer` to the `array of it's signatures`.
and the same with ` gg.Sign(creater.Account);`.
but transaction still fails with `Transaction signature verification `

**Comment by Neo-vortex** [comment]
> > I still don't understand why this example does not work
> > ```cs
> >   var tbuilder1 = new TransactionBuilder()
> >                 .SetRecentBlockHash((await rpcClient.GetRecentBlockHashAsync()).Result.Value.Blockhash)
> >                 .SetFeePayer(buyer.Account.PublicKey)
> >                 .AddInstruction(MemoProgram.NewMemo(creater.Account, "hello world"));
> >             var binmessge2  = tbuilder1.CompileMessage();
> >             var gg = Transaction.Populate(Message.Deserialize(binmessge2));
> >             gg.Sign(buyer.Account);
> >             gg.Sign(creater.Account);
> >  RequestResult<string> tx = await rpcClient.SendTransactionAsync(gg.Serialize());    //Transaction signature verification failure 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > As you said 'A transaction is a message along with the compact array of it's signatures.'. when I populate a transaction from a message with no signature , it means i am creating a transaction with empty signature array. so when i do `gg.Sign(buyer.Account);`, I am adding the signature of `buyer` to the `array of it's signatures`. and the same with ` gg.Sign(creater.Account);`. but transaction still fails with `Transaction signature verification `
> 
> In this example what does `gg.Sign(buyer.Account)` return? Internally that method also calls `VerifySignatures()` which verifies that all of the signatures present are valid, this is just to make sure what's going on. I'm going to attempt to replicate this.

This is the full example 0-100 : 
```csharp
            IRpcClient rpcClient2 = ClientFactory.GetClient(Cluster.TestNet);
            var wallet1 = new Wallet(WordCount.Fifteen, WordList.English);
            var wallet2 =  new Wallet(WordCount.Fifteen, WordList.English);
            var tbuilder1 = new TransactionBuilder()
                .SetRecentBlockHash((await rpcClient2.GetRecentBlockHashAsync()).Result.Value.Blockhash)
                .SetFeePayer(wallet1.Account.PublicKey)
                .AddInstruction(MemoProgram.NewMemo(wallet2.Account, "hello world"));
            var binmessge2  = tbuilder1.CompileMessage();
            var gg = Transaction.Populate(Message.Deserialize(binmessge2));
            var result1 =     gg.Sign(wallet1.Account);
            var result2 =   gg.Sign(wallet2.Account);
            RequestResult<string> txxx = await rpcClient2.SendTransactionAsync(gg.Serialize());    //Transaction signature verification failure 
            
            ```
```result1 = true```
```result2=true```

**Comment by murlokito** [comment]
> > > I still don't understand why this example does not work
> > > ```cs
> > >   var tbuilder1 = new TransactionBuilder()
> > >                 .SetRecentBlockHash((await rpcClient.GetRecentBlockHashAsync()).Result.Value.Blockhash)
> > >                 .SetFeePayer(buyer.Account.PublicKey)
> > >                 .AddInstruction(MemoProgram.NewMemo(creater.Account, "hello world"));
> > >             var binmessge2  = tbuilder1.CompileMessage();
> > >             var gg = Transaction.Populate(Message.Deserialize(binmessge2));
> > >             gg.Sign(buyer.Account);
> > >             gg.Sign(creater.Account);
> > >  RequestResult<string> tx = await rpcClient.SendTransactionAsync(gg.Serialize());    //Transaction signature verification failure 
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > As you said 'A transaction is a message along with the compact array of it's signatures.'. when I populate a transaction from a message with no signature , it means i am creating a transaction with empty signature array. so when i do `gg.Sign(buyer.Account);`, I am adding the signature of `buyer` to the `array of it's signatures`. and the same with ` gg.Sign(creater.Account);`. but transaction still fails with `Transaction signature verification `
> > 
> > 
> > In this example what does `gg.Sign(buyer.Account)` return? Internally that method also calls `VerifySignatures()` which verifies that all of the signatures present are valid, this is just to make sure what's going on. I'm going to attempt to replicate this.
> 
> This is the full example 0-100 :
> 
> ```cs
>             IRpcClient rpcClient2 = ClientFactory.GetClient(Cluster.TestNet);
>             var wallet1 = new Wallet(WordCount.Fifteen, WordList.English);
>             var wallet2 =  new Wallet(WordCount.Fifteen, WordList.English);
>             var tbuilder1 = new TransactionBuilder()
>                 .SetRecentBlockHash((await rpcClient2.GetRecentBlockHashAsync()).Result.Value.Blockhash)
>                 .SetFeePayer(wallet1.Account.PublicKey)
>                 .AddInstruction(MemoProgram.NewMemo(wallet2.Account, "hello world"));
>             var binmessge2  = tbuilder1.CompileMessage();
>             var gg = Transaction.Populate(Message.Deserialize(binmessge2));
>             var result1 =     gg.Sign(wallet1.Account);
>             var result2 =   gg.Sign(wallet2.Account);
>             RequestResult<string> txxx = await rpcClient2.SendTransactionAsync(gg.Serialize());    //Transaction signature verification failure         
> ```
> 
> `result1 = true` `result2=true`

We've replicated this and narrowed it down to an edge case in the `Transaction.Populate` method. A fix for this will be merged shortly and v0.5.1 will be released. Thanks for the help narrowing this down.
